Contracts/Presale/PresaleContract.sol

```diff
@@ -23,88 +23,95 @@ contract PresaleContract is WithdrawContract {
         uint totalTokens = 0;
         
-        while (deposit > 0 && index < 3) {
-            uint ethers = 0;
-            uint maxAmount = deposit * MINIMUM_TOKEN_VALUE_MULTIPLIER * prices[index] / 1 ether;
-            if (amounts[index] >= maxAmount) {
-                amounts[index] -= maxAmount;
-                ethers = deposit;
-                deposit = 0;
+        while (totalDeposit > 0 && stage < 3) {
+            uint currentStageDeposit = 0;
+            
+            uint maxTokensAmountForCurrentStage = totalDeposit * amountBoughtInOneWei[stage];
+            //checking free tokens amount for current stage
+            if (amountsLeft[stage] >= maxTokensAmountForCurrentStage) {
+                amountsLeft[stage] -= maxTokensAmountForCurrentStage;
+                currentStageDeposit = totalDeposit;
+                totalDeposit = 0;
             } else {
-                maxAmount = amounts[index];
-                amounts[index] = 0;
-                ethers = maxAmount * 1 ether / MINIMUM_TOKEN_VALUE_MULTIPLIER / prices[index];
-                deposit -= ethers;
+                maxTokensAmountForCurrentStage = amountsLeft[stage];
+                amountsLeft[stage] = 0;
+                currentStageDeposit = maxTokensAmountForCurrentStage / amountBoughtInOneWei[stage];
+                //if we have less tokens then min price we sell them for 1 wei
+                //example we have 101000 / 10^18 tokens 
+                //current price is 2500 tokens for 1 ether => 2500 / 10^18 tokens for 1 wei
+                //(101000 / 10^18 % (2500 / 10^18)) => (101000 % 2500) => 1000
+                //1000 > 0 so we sell 1000 / 10^18 tokens for 1 wei (minimum amount of ethers)
+                if ((maxTokensAmountForCurrentStage % amountBoughtInOneWei[stage]) > 0) {
+                    currentStageDeposit += 1;
+                }
+                totalDeposit -= currentStageDeposit;
                 
             }
             //save transaction
-            _addTransaction(msg.sender, ethers, prices[index], maxAmount);
+            _addTransaction(msg.sender, currentStageDeposit, amountBoughtInOneWei[stage], maxTokensAmountForCurrentStage);
             //calculations
-            totalTokens += maxAmount;
-            profit += ethers;
+            totalTokens += maxTokensAmountForCurrentStage;
+            currentInvestment += currentStageDeposit;
             //
-            index++;
+            stage++;
         }
         
-        _updateData(msg.sender, profit, totalTokens);
+        _updateUserBalance(msg.sender, currentInvestment, totalTokens);
         //return money
-        if (deposit > 0) {
-            withdraw(msg.sender, deposit);
+        if (totalDeposit > 0) {
+            withdraw(msg.sender, totalDeposit);
         }
         //auto-transfer to our wallet
-        if (profit > 0) {
-            withdraw(mainVallet, profit);
-            // mainVallet.transfer(profit);
+        if (currentInvestment > 0) {

```

Comment:
@Yaschenko please put these lines of transfer *above* the money return for the user